### PR TITLE
Cycle in Linked List 

### PR DESCRIPTION
**Pull request by dev24il**
**# Issue 415**
Cycle in Linked list #415 

# Cycle in Linked List
Given a linked list, check if the linked list has cycle or not. Below diagram shows a linked list with a cycle.  
image 

# Terms Explained:
Linked List : A linked list is a linear data structure, in which the elements are not stored at contiguous memory locations. 
image 

**Proposed Changes** 
- CycleDetection.cpp file added
- README.md file updated
[README.md](https://github.com/Lakhankumawat/LearnCPP/files/8281620/README.md)

Checklist
- [x] ✅ My code follows the code style of this project.
- [ ] 📝 My change requires a change to the README.
- [x] 🎀 I have updated the README accordingly.
- [x] 👀 I have read the [CONTRIBUTING](https://github.com/Lakhankumawat/LearnCPP/blob/main/Contributing.md) document.
- [x] ✨ I have added tests to cover my changes.
- [x] 🚩 All new and existing tests passed.
- [ ] 🌟 ed the repo
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

# Approach used: Floyd's algorithm
**This is the fastest method to detect cycle in linked list**
- Traverse linked list by using two pointers.
- Move one pointer(slow) by one step and second pointer(fast) by two step.
- If these pointers meet at the same node then there is a cycle. If pointers do not meet then linked list doesn’t have a cycle present.

## Algorithm
```
bool detectCycle(node* &head){
    node *slow = head;     
    node *fast = head;
    while(fast!=NULL && fast->next!=NULL){
        slow = slow->next;       
        fast = fast->next->next; 

        if(fast == slow){    
            return true;
        }
    }
    return false;    
}
```

# Properties
**Complexity Analysis:**
**Time Complexity: O(n)**
Only one traversal of the loop is needed.
**Space Complexity: O(1)**
There is no space required.

# **OUTPUT SCRRENSHOTS:**
**Title - User input of linked list to detect cycle**
Screenshot (492)

### CODE REVIEW

**Review by Lakhankumawat — CHANGES_REQUESTED** [review]
You need to follow particular guidelines for your README , please follow
these guidelines [click here](https://github.com/Lakhankumawat/LearnCPP#to-add-programs-using-stl-follow-same-approach-as--algorithms) 
also you can see how your README should look like [see here](https://github.com/Lakhankumawat/LearnCPP/blob/main/S-SortingAlgorithms/README.md) 
for more please go through these [here](https://youtu.be/7UgC3HDETRA)



Change Directory for code and README to LearnCPP/S-SinglyLinkedList/CycleDetection.cpp and README.md

**Comment by Lakhankumawat** [comment]
It is great you just skip all the checks

**Comment by Lakhankumawat** [comment]
Let me show you how to check the check list , there should be no space inside brackets like [x]
 `[x]`
- [x] Checklist
- [ ] Or you can just check them from outside without doing any of these
- [x] ✨ I have added tests to cover my changes.
- [x] 🚩 All new and existing tests passed.
- [ ] 🌟 ed the repo

1. Starr the repo if you haven't !
2. Have you watched this [video](https://youtu.be/kIFwqp9B7pQ) ?

File: L-LinkedList/S-SinglyLinkedList/CycleDetection/CycleDetection.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    // Linked list
+    insertAtTail(head, 1);
+    insertAtTail(head, 2);
+    insertAtTail(head, 3);
+    insertAtTail(head, 4);
+    insertAtTail(head, 5);
+    insertAtTail(head, 6);
+    makeCycle(head, 4);      
+
+    bool detect = detectCycle(head);
```
Comment (miraj0507) [inline L-LinkedList/S-SinglyLinkedList/CycleDetection/CycleDetection.cpp:77]: Kindly update this to user input style and not the hardcoded way of coding.

File: L-LinkedList/S-SinglyLinkedList/CycleDetection/CycleDetection.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    while(temp->next != NULL){
+        if(count == pos){
+            startNode = temp;
+        }
+        temp = temp->next;
+        count++;
+    }
+    temp->next = startNode;
+}
+
```
Comment (miraj0507) [inline L-LinkedList/S-SinglyLinkedList/CycleDetection/CycleDetection.cpp:48]: Why did you keep these two functions when its just cycle detection?

**Review by miraj0507 — COMMENTED** [review]
Kindly update the checklist as [x]

File: L-LinkedList/S-SinglyLinkedList/CycleDetection/CycleDetection.cpp (hunk 2 above)
Comment (dev24il) [reply L-LinkedList/S-SinglyLinkedList/CycleDetection/CycleDetection.cpp:48]: ok

**Comment by miraj0507** [comment]
This looks good to me, let's proceed further

File: L-LinkedList/S-SinglyLinkedList/CycleDetection/CycleDetection.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+// function to take user input of linked list
+node* addToList(node *head){
+    bool quit = false;
+    int temp;
+    node *current = new node(-1);
+
+    while (!quit){
+        cin >> temp;
+        if (temp == -1){
```
Comment (n4i9kita) [inline L-LinkedList/S-SinglyLinkedList/CycleDetection/CycleDetection.cpp:25]: I don't understand this `if` statement. Why would a user input -1? 
I think you should write some comment, or through output statement.

File: L-LinkedList/S-SinglyLinkedList/CycleDetection/CycleDetection.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            this->data = data;
+            next = NULL;
+        }
+};
+
+// function to take user input of linked list
+node* addToList(node *head){
+    bool quit = false;
+    int temp;
+    node *current = new node(-1);
```
Comment (n4i9kita) [inline L-LinkedList/S-SinglyLinkedList/CycleDetection/CycleDetection.cpp:21]: You should declare the `current` node in the `else` block. What if the user inputs -1? 
Will there be any use of memory allocation in that case?

File: L-LinkedList/S-SinglyLinkedList/CycleDetection/CycleDetection.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+    while (!quit){
+        cin >> temp;
+        if (temp == -1){
+            quit = true;
+        }
+        else{
+            current->data = temp;
+            current -> next = current;
```
Comment (n4i9kita) [inline L-LinkedList/S-SinglyLinkedList/CycleDetection/CycleDetection.cpp:30]: Please explain this to me.

`current -> next = current;` I am not able to understand it. Is it not a self loop?
How are you linking the node `currend` to the list for which we know `head`?

File: L-LinkedList/S-SinglyLinkedList/CycleDetection/README.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+```
+
+
+# Properties
+## Complexity Analysis:  
+## Time Complexity: O(n)
+- Only one traversal of the loop is needed.
+
+## Space Complexity: O(1)
+- There is no space required.
```
Comment (n4i9kita) [inline L-LinkedList/S-SinglyLinkedList/CycleDetection/README.md:64]: Please be careful with what you write.

File: L-LinkedList/S-SinglyLinkedList/CycleDetection/README.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+
+# Advantages
+- This is the fastest method to detect cycle in linked list
```
Comment (n4i9kita) [inline L-LinkedList/S-SinglyLinkedList/CycleDetection/README.md:68]: Line no. 26 and 68 are same.

File: L-LinkedList/S-SinglyLinkedList/CycleDetection/README.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+- There is no space required.
+
+
+# Disadvantages
+- Use of Two pointers
```
Comment (n4i9kita) [inline L-LinkedList/S-SinglyLinkedList/CycleDetection/README.md:73]: I think you should explain this.
What is the issue if I use two pointers?

**Pull request closed without merging**